### DISCUSSION TITLE
Getting an error while retrieving tokens

### DISCUSSION THREAD

**Discussion by remigauthierdocaposte** [InseeFr/Keycloak-FranceConnect#58] [Q&A]
Hello!

On our side, the external IDP is configured, using the specific nonce to be able to connect to FC+.

All good except when keycloak asks for the token using the returned authorizationCode collected after successfully calling the authorize.

I have got this error : 
`Failed to make identity provider oauth callback: org.keycloak.broker.provider.IdentityBrokerException: No access_token from server. error='invalid_client_metadata', error_description='client JSON Web Key Set failed to be refreshed', error_uri='null'`

Error seems to be located in the authResponse() of OIDCEndpoint here :
`SimpleHttp tokenRequest = SimpleHttp.doPost(tokenUrl, this.session).param("code", authorizationCode).param("redirect_uri", redirectUri).param("grant_type", "authorization_code");
SimpleHttp authenticateTokenRequest = FranceConnectIdentityProvider.this.authenticateTokenRequest(tokenRequest);`

Do you have any clue or advice? We currently only have an integration client_id/client_secret for FC+.

**Comment by micedre** [comment]
This extension is currently not compatible with FC+, I'm waiting for a working client_id to see how much should be changed (if it is possible) to make it working. 

Could you maybe share all keycloak logs (removing all sensitive informations) by activating http client wire logging ?

(To enable wire logging, open jboss-cli on your keycloak server `$KEYCLOAK_HOME/bin/jboss-cli.sh` and then : 
```
> connect
>/subsystem=logging/logger=org.apache.http.wire:add(handlers=[CONSOLE,FILE], level=TRACE)
```
You should see all http exchange from keycloak server with the FC+ server in server.log file (not in console)

**Reply by remigauthierdocaposte** [reply]
Thanks for you answer, I ll try that first hour next week and keep you posted.

**Reply by remigauthierdocaposte** [reply]
It seems FranceConnect improved the message on their side since 2 days ago, it is more clear now : "{"error":"invalid_client_metadata","error_description":"no suitable encryption key found (RSA-OAEP)"}"
I ll try to add an RSA-OAEP key as specified in their documentation with type "enc".
Not sure keycloak will allow me to do this easily...

**Reply by remigauthierdocaposte** [reply]
https://issues.redhat.com/browse/KEYCLOAK-19318

**Reply by micedre** [reply]
Ok I added a provider to add an encryption key to jwks uri, like this : https://keycloak.dev.insee.io/auth/realms/test-fc/protocol/openid-connect/certs
I'll try to generate a good type of key with this.

**Reply by remigauthierdocaposte** [reply]
Did you manage to add an RSA-OAEP enc key?
_I ll try to add an enc key of the good type using the Admin API with Postman with a PUT to mimic an update on an existing rsa-generated sig key._ (Edit : not working / Edit2 : finally working using service account with realm rights)

Also do you manage to see the enc key in the jwks uri?
There is a bug on this. Which provider are you using to be able to see this key?

**Reply by remigauthierdocaposte** [reply]
Ok I managed to force add an RSA-OAEP-256 and RSA-OAEP using Admin API and a service account.

![image](https://user-images.githubusercontent.com/89779307/139848249-d25b0149-1e7a-4c35-a37e-8eff99d80269.png)

Not better with RSA-OAEP-256 but I got a response token using RSA-OAEP as prioritary enc alg.

![image](https://user-images.githubusercontent.com/89779307/139851397-494617f4-28b3-42fc-a6e1-d680fee47e1a.png)

**Reply by remigauthierdocaposte** [reply]
To add such a key use a service account with realm rights.
Call POST https://URL/auth/admin/realms/REALM/components with bearer access token of service account and add raw json to the body with `{"name":"rsa-oaep-franceconnectplus","providerId":"rsa-generated","providerType":"org.keycloak.keys.KeyProvider","parentId":REALM",
"config":{"priority":["200"],"enabled":["true"],"active":["true"],"algorithm":["RSA-OAEP"],"keyUse":["enc"]}}`

**Reply by remigauthierdocaposte** [reply]
I will try to use my key to get the ID Token information from the returned JWT.
I you need more info to do tests on your side, so not hesitate to send me a mail.

**Reply by micedre** [reply]
Thank you for all that, we will try on our side

**Reply by micedre** [reply]
I finally had some way to test the service, could you try the extension here : https://minio.lab.sspcloud.fr/projet-keycloak-franceconnect/keycloak-franceconnect-2.4-SNAPSHOT.jar

It works for me.

It still needs a lot of code cleaning.

**Reply by remigauthierdocaposte** [reply]
Ok thanks for the share, I ll sure try it ASAP and let you know the results of our tests.

**Reply by remigauthierdocaposte** [reply]
Good news it is working fine!! I ll have a look at your code to check the details of your implementation. Excellent work. Thanks again for the talk.
